lib/Dialect/FIRRTL/Transforms/PassiveWires.cpp

```diff
@@ -0,0 +1,81 @@
+//===- PassiveWires.cpp - Make Wires Passive --------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the PassiveWires pass.  This pass eliminated flips from
+// wires with aggregate types.  Since flips only determine connect direction,
+// they are unnecessary on wires and just get in the way.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-passive-wires"
+
+using namespace circt;
+using namespace firrtl;
+
+static bool hasFlip(Type t) {
+  if (auto type = type_dyn_cast<FIRRTLBaseType>(t))
+    return !type.isPassive();
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct PassiveWiresPass : public PassiveWiresBase<PassiveWiresPass> {
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+// This is the main entrypoint for the lowering pass.
+void PassiveWiresPass::runOnOperation() {
+  LLVM_DEBUG(
+      llvm::dbgs() << "===- Running Passive Wires Pass "
+                      "------------------------------------------------===\n");
+  auto module = getOperation();
+
+  // First, expand any connects to resolve flips
+  SmallVector<WireOp> wires;
+  module.walk([&](Operation *op) -> WalkResult {
+    if (auto wire = dyn_cast<WireOp>(op)) {
+      if (hasFlip(wire.getType(0)))
+        wires.push_back(wire);
+      return WalkResult::advance();
+    }
+    if (!isa<ConnectOp, StrictConnectOp>(op))
+      return WalkResult::advance();
+    // connect/strictconnect
+    if (!hasFlip(op->getOperand(0).getType()))
+      return WalkResult::advance();
+
+    mlir::ImplicitLocOpBuilder builder(op->getLoc(), op);
+    // This will "blow out" a connect to passive pieces
+    emitConnect(builder, op->getOperand(0), op->getOperand(1));
+    op->erase();
+    return WalkResult::advance();
+  });
+
+  // Second, remove flips from wires.
+  for (auto w : wires) {
+    auto r = w.getResult();
+    // in-place updates is safe as consumers don't care about flip
+    r.setType(type_cast<FIRRTLBaseType>(r.getType()).getPassiveType());
+  }
```

Comment:
I think you can use a reverse iterator to avoid the iteration over the wires. I.e., you are then guaranteed to walk the connects before the wires so you can rewrite the wires in the same walk.

Reply:
I don't think there's an easy way to do a `Operation::walk` in reverse order 🤔

Reply:
I think the template for `mlir::Operation::walk` takes as its second argument that should let you control whether it is forward (default) or reverse. The documentation seems to support this: https://mlir.llvm.org/doxygen/classmlir_1_1Operation.html#a59740592240b950b8c8afcf4a2eb4113

(I've never tried this!)

Reply:
Oh that's cool! TIL: walk supports reverse iteration 🤯

Reply:
You can go pre or post order over the nesting, but only forward iteration order seems to exist upstream.